src/Bicep.Core/Semantics/Namespaces/SystemNamespaceSymbol.cs

```diff
@@ -508,6 +514,50 @@ private static SyntaxBase StringLiteralFunctionReturnTypeEvaluator(FunctionCallS
             return SyntaxFactory.CreateStringLiteral(stringLiteral.RawStringValue);
         }
 
+        private static TypeSymbol GetItemsReturnType(TypeSymbol keyType, TypeSymbol valueType)
+            => new TypedArrayType(
+                new ObjectType(
+                    "object",
+                    TypeSymbolValidationFlags.Default,
+                    new [] {
+                        new TypeProperty("key", keyType, description: "The key of the object property being iterated over."),
+                        new TypeProperty("value", valueType, description: "The value of the object property being iterated over."),
+                    },
+                    null),
+                TypeSymbolValidationFlags.Default);
+
+        private static TypeSymbol ItemsTypeBuilder(IBinder binder, IFileResolver fileResolver, IDiagnosticWriter diagnostics, ImmutableArray<FunctionArgumentSyntax> arguments, ImmutableArray<TypeSymbol> argumentTypes)
+        {
+            if (argumentTypes[0] is not ObjectType objectType)
+            {
+                return GetItemsReturnType(LanguageConstants.String, LanguageConstants.Any);
+            }
+
+            var keyTypes = new List<TypeSymbol>();
+            var valueTypes = new List<TypeSymbol>();
+            foreach (var property in objectType.Properties.Values)
+            {
+                if (property.Flags.HasFlag(TypePropertyFlags.WriteOnly))
+                {
+                    // we're reading the values - exclude write-only properties
+                    continue;
+                }
+
+                keyTypes.Add(new StringLiteralType(property.Name));
+                valueTypes.Add(property.TypeReference.Type);
+            }
+
+            if (objectType.AdditionalPropertiesType?.Type is {} additionalPropertiesType)
+            {
+                keyTypes.Add(LanguageConstants.String);
+                valueTypes.Add(additionalPropertiesType);
+            }
+
+            return GetItemsReturnType(
+                keyType: UnionType.Create(keyTypes),
+                valueType: UnionType.Create(valueTypes));
```

Comment:
>Create [](http://example.com/codeflow?start=37&length=6)

If you have something like this, we will get a type error:

```bicep
var o = {
  a: 'hello'
  b: 42
}

var a = items(o)

resource r '...' = {
  name: 'name'
  properties: {
    '<some int property>': a[1].value // error in this assignment
  }
}
```

We can't avoid it until we implement support in the type system for having different types for items by index on array literals. Do we just live with the limitation because it's a niche case but get better completions in more common cases? Or do we sacrifice some of the completions and hovers and return something more generic like an array of objects where `key` and `value` are of `any` type?

Reply:
Good point, I agree. I've been meaning to look at the implementation for the code which converts an array of ints to `int[]` and an array of mixed items to `array` - feels like perhaps I should use that same implementation here.